### PR TITLE
[skip ci] Add examples for has_{one,many} :through :source and :source_type

### PR DESCRIPTION
**Pull request by imtayadeway**
### Summary
I've often been confused about `source` and `source_type`. I thought it would be helpful to provide some examples here.

### CODE REVIEW

File: guides/source/association_basics.md [hunk 1]
```diff
@@ -1301,10 +1301,40 @@ By convention, Rails assumes that the column used to hold the primary key of thi
...
+class Jpeg < ApplicationRecord; end
+class Png < ApplicationRecord; end
+
+class Profile < ApplicationRecord
+  has_one :picture, polymorphic: true
+end
+
+class User < ApplicationRecord
+  has_one :profile
+  has_one :jpeg, through: :profile, source: :picture, source_type: "Jpeg"
```
Comment (imtayadeway) [inline guides/source/association_basics.md]: FWIW, I do not love this example, but really struggled to come up with one for a polymorphic has_one :through. Will gladly take suggestions for a better one!

Comment (vipulnsward) [reply guides/source/association_basics.md]: Can we combine the source and source_type examples together? Lots of repetition in both.

File: guides/source/association_basics.md [hunk 2]
```diff
@@ -1713,10 +1743,41 @@ record's `user_id` value will be the `guid` value of `@user`.
...
+  belongs_to :assembly
+  belongs_to :part
+end
+
+class Part < ApplicationRecord
+  has_many :manifests
+  has_many :assemblies, through: :manifests
+end
+```
+
```
Comment (vipulnsward) [inline guides/source/association_basics.md]: Same here, we can combine the source/type examples into one and just show below source_type

**Comment by imtayadeway** [comment]
@vipulnsward removed the `source`-only examples (as `source` is used in all the examples here) as suggested and updated the example in https://github.com/rails/rails/pull/35612/commits/c9e9feb37e488376f11cbd148401f78445d39e84 to something perhaps a bit more realistic. LMK if you have any other suggestions. Thanks!

File: guides/source/association_basics.md [hunk 3]
```diff
@@ -1305,6 +1305,20 @@ The `:source` option specifies the source association name for a `has_one :throu
...
+class Employment < ApplicationRecord
+  has_one :health_care_plan, polymorphic: true
+end
+
+class User < ApplicationRecord
+  has_one :employment
+  has_one :hdhp, through: :employment, source: :health_care_plan, source_type: "Hdhp"
+end
+```
```
Comment (eileencodes) [inline guides/source/association_basics.md]: Can you use a different example? I think that health care plan is confusing for non-US readers. Even US folks may not understand the difference between PPO and high deductible plan. I have one and had to look up "HDHP". :smile:

**Review by eileencodes — CHANGES_REQUESTED** [review]
Reading this over I think it's a good idea to document but we should use the same example for all the docs. In this file we use accounts, books, and now health insurance plans and doctor specialities. I think we should use books - it's simple to understand and translate and doesn't rely on understanding differences of our health insurance systems.

File: guides/source/association_basics.md (hunk 3 above)
Comment (imtayadeway) [reply guides/source/association_basics.md]: @eileencodes I'd be really happy to. My first thought was to extend one of the existing example domains, but I had the hardest time coming up with a polymorphic has_one :through example that made sense. I totally agree that the one I settled on is not very obvious/accessible. I'll give this some more thought, though I'd happily take a suggestion!

**Comment by imtayadeway** [comment]
@eileencodes thanks for your patience! I updated both examples to use the books domain - I hope that they make sense! Happy to accommodate any further suggestions :smile:

**Review by eileencodes — APPROVED** [review]
I like these examples much better thank you!

**Pull request merged**